kernel/src/transaction.rs

```diff
@@ -0,0 +1,253 @@
+        Transaction {
+            read_snapshot: snapshot.into(),
+            operation: None,
+            commit_info: None,
+        }
+    }
+
+    /// Consume the transaction and commit the in-progress write to the table.
+    pub fn commit(self, engine: &dyn Engine) -> DeltaResult<CommitResult> {
+        // step one: construct the iterator of actions we want to commit
+        // note: only support commit_info right now.
+        let actions: Box<dyn Iterator<Item = Box<dyn EngineData>> + Send> = match self.commit_info {
+            Some(engine_commit_info) => {
+                let actions =
+                    generate_commit_info(engine, self.operation.as_deref(), engine_commit_info)?;
+                Box::new(std::iter::once(actions))
+            }
+            None => {
+                // if there is no commit info, actions start empty
+                Box::new(std::iter::empty())
+            }
+        };
+
+        // step two: set new commit version (current_version + 1) and path to write
+        let commit_version = self.read_snapshot.version() + 1;
+        let commit_path =
+            ParsedLogPath::new_commit(self.read_snapshot.table_root(), commit_version)?;
+
+        // step three: commit the actions as a json file in the log
+        let json_handler = engine.get_json_handler();
+        match json_handler.write_json_file(&commit_path.location, Box::new(actions), false) {
+            Ok(()) => Ok(CommitResult::Committed(commit_version)),
+            Err(Error::FileAlreadyExists(_)) => Err(Error::generic("fixme")), // Ok(CommitResult::Conflict(self, commit_version)),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Set the operation that this transaction is performing. This string will be persisted in the
+    /// commitInfo action in the log. <- TODO include this bit?
+    pub fn operation(&mut self, operation: String) {
+        self.operation = Some(operation);
+    }
+
+    /// WARNING: This is an unstable API and will likely change in the future.
+    ///
+    /// Add commit info to the transaction. This is commit-wide metadata that is written as the
+    /// first action in the commit. The engine data passed here must have exactly one row, and we
+    /// only read one column: `engineCommitInfo` which must be a map<string, string> encoding the
+    /// metadata.
+    ///
+    /// Note that there are two main pieces of information included in commit info: (1) custom
+    /// engine commit info (specified via this API) and (2) delta's own commit info which is
+    /// effectively appended to the engine-specific commit info.
+    ///
+    /// If the engine elects to omit commit info, it can do so in two ways:
+    /// 1. skip this API entirely - this will omit the commitInfo action from the commit (that is,
+    ///    it will prevent the kernel from writing delta-specific commitInfo). This precludes usage
+    ///    of table features which require commitInfo like in-commit timestamps.
```

Comment:
yup - i now require commit_info to be present. and I've clarified the comment, though for now we still disallow having a column other than `engineCommitInfo` (that is, we require their engine data chunk has an `engineCommitInfo` column) otherwise the input to expression evaluation has an unknown schema.